signalflow/messages/binary.go

```diff
@@ -0,0 +1,184 @@
+package messages
+
+import (
+	"bytes"
+	"compress/gzip"
+	"encoding/binary"
+	"fmt"
+	"io/ioutil"
+	"math"
+
+	"github.com/signalfx/signalfx-go/idtool"
+)
+
+type BinaryPayload struct {
+	Type ValType
+	TSID idtool.ID
+	Val  [8]byte
+}
+
+// Value returns the numeric value as an interface{}.
+func (bp *BinaryPayload) Value() interface{} {
+	switch bp.Type {
+	case ValTypeLong:
+		return bp.Int64()
+	case ValTypeDouble:
+		return bp.Float64()
+	case ValTypeInt:
+		return bp.Int32()
+	default:
+		return nil
+	}
+}
+
+func (bp *BinaryPayload) Int64() int64 {
+	n := binary.BigEndian.Uint64(bp.Val[:])
+	return int64(n)
+}
+
+func (bp *BinaryPayload) Float64() float64 {
+	bits := binary.BigEndian.Uint64(bp.Val[:])
+	return math.Float64frombits(bits)
+}
+
+func (bp *BinaryPayload) Int32() int32 {
+	var n int32
+	_ = binary.Read(bytes.NewBuffer(bp.Val[:]), binary.BigEndian, &n)
+	return n
+}
+
+// DataMessage is a set of datapoints that share a common timestamp
+type DataMessage struct {
+	BaseMessage
+	BaseChannelMessage
+	TimestampedMessage
+	Payloads []BinaryPayload
+}
+
+func (dm *DataMessage) String() string {
+	pls := make([]map[string]interface{}, 0)
+	for _, pl := range dm.Payloads {
+		pls = append(pls, map[string]interface{}{
+			"type":  pl.Type,
+			"tsid":  pl.TSID,
+			"value": pl.Value(),
+		})
+	}
+
+	return fmt.Sprintf("%v", map[string]interface{}{
+		"channel":   dm.Channel(),
+		"timestamp": dm.Timestamp(),
+		"payloads":  pls,
+	})
+}
+
+type binaryMessageHeader struct {
+	TimestampMillis uint64
+	ElementCount    uint32
+}
+
+type ValType uint8
+
+const (
+	ValTypeLong   ValType = 1
+	ValTypeDouble ValType = 2
+	ValTypeInt    ValType = 3
+)
+
+func (vt ValType) String() string {
+	switch vt {
+	case ValTypeLong:
+		return "long"
+	case ValTypeDouble:
+		return "double"
+	case ValTypeInt:
+		return "int32"
+	}
+	return "Unknown"
+}
+
+// The first 20 bytes of every binary websocket message from the backend.
+// https://developers.signalfx.com/signalflow_analytics/rest_api_messages/stream_messages_specification.html#_binary_encoding_of_websocket_messages
+type msgHeader struct {
+	Version     uint8
+	MessageType uint8
+	Flags       uint8
+	Reserved    uint8
+	Channel     [16]byte
+}
+
+const (
+	compressed  uint8 = 1 << iota
+	jsonEncoded       = 1 << iota
+)
+
+func parseBinaryHeader(msg []byte) (string, bool /* isCompressed */, bool /* isJSON */, []byte /* rest of message */, error) {
```

Comment:
Could name the return values instead of doing it as comments.

Reply:
I prefer we don't use named returns due to shadowing and the surprise they cause. (Since in the body of the function, there's no way a new developer knows that variable will be returned without reading the function definition.)

The documentation is a big weird, but I'd rather have that than named returns. (I dunno if we have any code guidelines internally for Go that I am arguing with, but am happy to have those chats and be converted if needed.)

Reply:
Generally the best thing is to make a struct type that holds the return values in named fields, but I was trying to avoid having another type just to represent the parsed binary header since we already have one for the binary reader's sake.  Yeah, I agree about named returns -- every time I encounter them they do more to make things harder to follow than help.